File: tests/unit/test_image.py
"""
Unit tests for image deduplication module.
"""

import sys
from pathlib import Path
from unittest import mock

import pytest
from PIL import Image

# Add src to path for imports
sys.path.insert(0, str(Path(__file__).parent.parent.parent / "src"))

from dedupl.image import (
    ImageMeta,
    compute_perceptual_hash,
    group_by_exact_hash,
    group_by_perceptual_hash,
    read_image_dimensions,
    select_best_quality,
)


class TestImageMeta:
    """Test ImageMeta dataclass."""

    def test_image_meta_creation(self, tmp_path):
        """Test creating ImageMeta instance."""
        test_file = tmp_path / "test.jpg"
        test_file.touch()

        meta = ImageMeta(
            path=test_file,
            width=1920,
            height=1080,
            size=2048000,
            mtime=1234567890.0,
            sha1="abc123def456",
            phash=None,
        )

        assert meta.path == test_file
        assert meta.width == 1920
        assert meta.height == 1080
        assert meta.size == 2048000
        assert meta.mtime == 1234567890.0
        assert meta.sha1 == "abc123def456"
        assert meta.phash is None

    def test_image_meta_immutable(self, tmp_path):
        """Test that ImageMeta is immutable."""
        test_file = tmp_path / "test.jpg"
        meta = ImageMeta(
            path=test_file,
            width=1920,
            height=1080,
            size=2048000,
            mtime=1234567890.0,
            sha1="abc123def456",
            phash=None,
        )

        with pytest.raises(AttributeError):
            meta.width = 3840


class TestImageDimensions:
    """Test image dimension reading."""

    def test_read_dimensions_success(self, tmp_path):
        """Test successful dimension reading."""
        test_file = tmp_path / "test.jpg"

        # Create a small test image
        img = Image.new("RGB", (100, 50), color="red")
        img.save(test_file)

        width, height = read_image_dimensions(test_file)
        assert width == 100
        assert height == 50

    def test_read_dimensions_failure(self, tmp_path):
        """Test dimension reading for invalid file."""
        test_file = tmp_path / "invalid.jpg"
        test_file.write_text("not an image")

        width, height = read_image_dimensions(test_file)
        assert width == 0
        assert height == 0

    def test_read_dimensions_nonexistent(self):
        """Test dimension reading for nonexistent file."""
        width, height = read_image_dimensions(Path("/nonexistent.jpg"))
        assert width == 0
        assert height == 0


class TestPerceptualHash:
    """Test perceptual hash computation."""

    def test_compute_perceptual_hash_success(self, tmp_path):
        """Test successful perceptual hash computation."""
        test_file = tmp_path / "test.jpg"

        # Create a test image
        img = Image.new("RGB", (100, 100), color="blue")
        img.save(test_file)

        phash = compute_perceptual_hash(test_file, hash_size=8)
        assert phash is not None

    def test_compute_perceptual_hash_different_images(self, tmp_path):
        """Test that different images have different hashes."""
        file1 = tmp_path / "img1.jpg"
        file2 = tmp_path / "img2.jpg"

        # Create different images
        img1 = Image.new("RGB", (100, 100), color="red")
        img1.save(file1)

        img2 = Image.new("RGB", (100, 100), color="blue")
        # Make the second image more different by adding a pattern
        pixels = img2.load()
        for i in range(0, 100, 10):
            for j in range(0, 100, 10):
                pixels[i, j] = (0, 0, 255)  # Blue
                pixels[i + 5, j + 5] = (255, 255, 0)  # Yellow
        img2.save(file2)

        hash1 = compute_perceptual_hash(file1, hash_size=8)
        hash2 = compute_perceptual_hash(file2, hash_size=8)

        # Compare hashes using Hamming distance
        distance = hash1 - hash2
        assert distance > 0  # Different images should have different hashes

    def test_compute_perceptual_hash_similar_images(self, tmp_path):
        """Test that similar images have close hashes."""
        file1 = tmp_path / "img1.jpg"
        file2 = tmp_path / "img2.jpg"

        # Create base image
        img1 = Image.new("RGB", (100, 100), color="red")
        img1.save(file1)

        # Create slightly modified image (small change)
        img2 = Image.new("RGB", (100, 100), color="red")
        # Add a small mark
        pixels = img2.load()
        pixels[50, 50] = (255, 0, 0)
        img2.save(file2)

        hash1 = compute_perceptual_hash(file1, hash_size=8)
        hash2 = compute_perceptual_hash(file2, hash_size=8)

        # Hashes should be very similar (small Hamming distance)
        distance = hash1 - hash2
        assert distance < 5


class TestQualitySelection:
    """Test image quality selection."""

    def test_select_best_quality_resolution_priority(self, tmp_path):
        """Test that higher resolution is prioritized."""
        high_res = ImageMeta(
            path=tmp_path / "high.jpg",
            width=3840,
            height=2160,
            size=4000000,
            mtime=1000.0,
            sha1="abc123",
            phash=None,
        )

        low_res = ImageMeta(
            path=tmp_path / "low.jpg",
            width=1920,
            height=1080,
            size=2000000,
            mtime=2000.0,  # Newer
            sha1="def456",
            phash=None,
        )

        best = select_best_quality([low_res, high_res])
        assert best == high_res

    def test_select_best_quality_size_tiebreaker(self, tmp_path):
        """Test that file size is used as tiebreaker."""
        file1 = ImageMeta(
            path=tmp_path / "file1.jpg",
            width=1920,
            height=1080,
            size=3000000,  # Larger
            mtime=1000.0,
            sha1="abc123",
            phash=None,
        )

        file2 = ImageMeta(
            path=tmp_path / "file2.jpg",
            width=1920,
            height=1080,
            size=2000000,
            mtime=1000.0,
            sha1="def456",
            phash=None,
        )

        best = select_best_quality([file1, file2])
        assert best == file1

    def test_select_best_quality_mtime_tiebreaker(self, tmp_path):
        """Test that modification time is used as tiebreaker."""
        file1 = ImageMeta(
            path=tmp_path / "file1.jpg",
            width=1920,
            height=1080,
            size=2000000,
            mtime=2000.0,  # Newer
            sha1="abc123",
            phash=None,
        )

        file2 = ImageMeta(
            path=tmp_path / "file2.jpg",
            width=1920,
            height=1080,
            size=2000000,
            mtime=1000.0,
            sha1="def456",
            phash=None,
        )

        best = select_best_quality([file1, file2])
        assert best == file1


class TestExactHashGrouping:
    """Test exact hash grouping."""

    def test_group_by_exact_hash_identical(self, tmp_path):
        """Test grouping identical files."""
        meta1 = ImageMeta(
            path=tmp_path / "file1.jpg",
            width=1920,
            height=1080,
            size=2000000,
            mtime=1000.0,
            sha1="abc123",
            phash=None,
        )

        meta2 = ImageMeta(
            path=tmp_path / "file2.jpg",
            width=1920,
            height=1080,
            size=2000000,
            mtime=1000.0,
            sha1="abc123",  # Same hash
            phash=None,
        )

        groups = group_by_exact_hash([meta1, meta2])
        assert len(groups) == 1
        assert len(groups[0]) == 2

    def test_group_by_exact_hash_different(self, tmp_path):
        """Test that different hashes don't group."""
        meta1 = ImageMeta(
            path=tmp_path / "file1.jpg",
            width=1920,
            height=1080,
            size=2000000,
            mtime=1000.0,
            sha1="abc123",
            phash=None,
        )

        meta2 = ImageMeta(
            path=tmp_path / "file2.jpg",
            width=1920,
            height=1080,
            size=2000000,
            mtime=1000.0,
            sha1="def456",  # Different hash
            phash=None,
        )

        groups = group_by_exact_hash([meta1, meta2])
        assert len(groups) == 0  # No groups with more than 1 item

    def test_group_by_exact_hash_multiple_groups(self, tmp_path):
        """Test multiple duplicate groups."""
        items = [
            ImageMeta(
                tmp_path / f"file{i}.jpg",
                100,
                100,
                1000,
                1000.0,
                sha1="hash1" if i < 2 else "hash2",
                phash=None,
            )
            for i in range(4)
        ]

        groups = group_by_exact_hash(items)
        assert len(groups) == 2  # Two groups
        assert all(len(group) == 2 for group in groups)


class TestPerceptualHashGrouping:
    """Test perceptual hash grouping."""

    @mock.patch("dedupl.image.imagehash.ImageHash")
    def test_group_by_perceptual_hash_similar(self, mock_hash_class, tmp_path):
        """Test grouping perceptually similar images."""
        # Create mock hashes with controlled distance
        hash1 = mock.Mock()
        hash2 = mock.Mock()
        hash1.__sub__ = mock.Mock(return_value=3)  # Distance of 3

        meta1 = ImageMeta(
            path=tmp_path / "file1.jpg",
            width=1920,
            height=1080,
            size=2000000,
            mtime=1000.0,
            sha1="abc123",
            phash=hash1,
        )

        meta2 = ImageMeta(
            path=tmp_path / "file2.jpg",
            width=1920,
            height=1080,
            size=2000000,
            mtime=1000.0,
            sha1="def456",
            phash=hash2,
        )

        groups = group_by_perceptual_hash([meta1, meta2], threshold=5)
        assert len(groups) == 1
        assert len(groups[0]) == 2

    @mock.patch("dedupl.image.imagehash.ImageHash")
    def test_group_by_perceptual_hash_different(self, mock_hash_class, tmp_path):
        """Test that dissimilar images don't group."""
        # Create mock hashes with large distance
        hash1 = mock.Mock()
        hash2 = mock.Mock()
        hash1.__sub__ = mock.Mock(return_value=20)  # Distance of 20

        meta1 = ImageMeta(
            path=tmp_path / "file1.jpg",
            width=1920,
            height=1080,
            size=2000000,
            mtime=1000.0,
            sha1="abc123",
            phash=hash1,
        )

        meta2 = ImageMeta(
            path=tmp_path / "file2.jpg",
            width=1920,
            height=1080,
            size=2000000,
            mtime=1000.0,
            sha1="def456",
            phash=hash2,
        )

        groups = group_by_perceptual_hash([meta1, meta2], threshold=5)
        assert len(groups) == 0  # No groups formed

    def test_group_by_perceptual_hash_aspect_ratio_check(self, tmp_path):
        """Test aspect ratio checking in perceptual grouping."""
        # Mock hashes with small distance
        hash1 = mock.Mock()
        hash2 = mock.Mock()
        hash1.__sub__ = mock.Mock(return_value=3)

        # Different aspect ratios
        meta1 = ImageMeta(
            path=tmp_path / "wide.jpg",
            width=1920,
            height=1080,  # 16:9
            size=2000000,
            mtime=1000.0,
            sha1="abc123",
            phash=hash1,
        )

        meta2 = ImageMeta(
            path=tmp_path / "square.jpg",
            width=1000,
            height=1000,  # 1:1
            size=2000000,
            mtime=1000.0,
            sha1="def456",
            phash=hash2,
        )

        # With aspect ratio check, shouldn't group
        groups = group_by_perceptual_hash([meta1, meta2], threshold=5, check_aspect_ratio=True)
        assert len(groups) == 0

        # Without aspect ratio check, should group
        groups = group_by_perceptual_hash([meta1, meta2], threshold=5, check_aspect_ratio=False)
        # This would group if hashes were actually similar

File: tests/test_image.rs
//! Integration tests for `image` module — parity with `test_image.py`.

use std::path::PathBuf;

use tempfile::TempDir;

use dedupl::image::grouping::{group_by_exact_hash, group_by_perceptual_hash};
use dedupl::image::meta::ImageMeta;
use dedupl::image::phash::{compute_perceptual_hash, hamming_distance, read_image_dimensions};
use dedupl::image::quality::select_best_quality;

// ── TestImageMeta ─────────────────────────────────────────────────────────

#[test]
fn test_image_meta_creation() {
    let meta = ImageMeta {
        path: PathBuf::from("/tmp/test.jpg"),
        width: 1920,
        height: 1080,
        size: 2_048_000,
        mtime: 1_234_567_890.0,
        sha1: "abc123def456".to_string(),
        phash: None,
    };

    assert_eq!(meta.width, 1920);
    assert_eq!(meta.height, 1080);
    assert_eq!(meta.size, 2_048_000);
    assert_eq!(meta.sha1, "abc123def456");
    assert!(meta.phash.is_none());
}

#[test]
fn test_image_meta_immutable_by_design() {
    // Rust enforces this; a non-`mut` binding cannot be modified.
    let meta = ImageMeta {
        path: PathBuf::from("/tmp/test.jpg"),
        width: 1920,
        height: 1080,
        size: 2_048_000,
        mtime: 1_234_567_890.0,
        sha1: "abc123def456".to_string(),
        phash: None,
    };
    assert_eq!(meta.width, 1920);
}

// ── TestImageDimensions ───────────────────────────────────────────────────

#[test]
fn test_read_dimensions_success() {
    let tmp = TempDir::new().unwrap();
    let test_file = tmp.path().join("test.png");

    // Create a 100×50 red image using the `image` crate.
    let img = image::RgbImage::from_fn(100, 50, |_, _| image::Rgb([255, 0, 0]));
    img.save(&test_file).unwrap();

    let (w, h) = read_image_dimensions(&test_file);
    assert_eq!(w, 100);
    assert_eq!(h, 50);
}

#[test]
fn test_read_dimensions_failure() {
    let tmp = TempDir::new().unwrap();
    let test_file = tmp.path().join("invalid.jpg");
    std::fs::write(&test_file, "not an image").unwrap();

    let (w, h) = read_image_dimensions(&test_file);
    assert_eq!(w, 0);
    assert_eq!(h, 0);
}

#[test]
fn test_read_dimensions_nonexistent() {
    let (w, h) = read_image_dimensions(std::path::Path::new("/nonexistent.jpg"));
    assert_eq!(w, 0);
    assert_eq!(h, 0);
}

// ── TestPerceptualHash ────────────────────────────────────────────────────

#[test]
fn test_compute_perceptual_hash_success() {
    let tmp = TempDir::new().unwrap();
    let test_file = tmp.path().join("test.png");

    let img = image::RgbImage::from_fn(100, 100, |_, _| image::Rgb([0u8, 0, 255]));
    img.save(&test_file).unwrap();

    let phash = compute_perceptual_hash(&test_file, 8);
    assert!(phash.is_some());
}

#[test]
fn test_compute_perceptual_hash_different_images() {
    let tmp = TempDir::new().unwrap();

    let file1 = tmp.path().join("img1.png");
    let img1 = image::RgbImage::from_fn(100, 100, |_, _| image::Rgb([255u8, 0, 0]));
    img1.save(&file1).unwrap();

    let file2 = tmp.path().join("img2.png");
    let img2 = image::RgbImage::from_fn(100, 100, |x, y| {
        if (x % 10 == 0) && (y % 10 == 0) {
            image::Rgb([255u8, 255, 0])
        } else {
            image::Rgb([0u8, 0, 255])
        }
    });
    img2.save(&file2).unwrap();

    let hash1 = compute_perceptual_hash(&file1, 8).unwrap();
    let hash2 = compute_perceptual_hash(&file2, 8).unwrap();

    let distance = hamming_distance(&hash1, &hash2);
    assert!(distance > 0, "Different images should have different hashes");
}

#[test]
fn test_compute_perceptual_hash_similar_images() {
    let tmp = TempDir::new().unwrap();

    let file1 = tmp.path().join("img1.png");
    let img1 = image::RgbImage::from_fn(100, 100, |_, _| image::Rgb([255u8, 0, 0]));
    img1.save(&file1).unwrap();

    // Nearly identical — one pixel changed.
    let file2 = tmp.path().join("img2.png");
    let img2 = image::RgbImage::from_fn(100, 100, |x, y| {
        if x == 50 && y == 50 {
            image::Rgb([254u8, 0, 0])
        } else {
            image::Rgb([255u8, 0, 0])
        }
    });
    img2.save(&file2).unwrap();

    let hash1 = compute_perceptual_hash(&file1, 8).unwrap();
    let hash2 = compute_perceptual_hash(&file2, 8).unwrap();

    let distance = hamming_distance(&hash1, &hash2);
    assert!(distance < 5, "Similar images should have close hashes, got {distance}");
}

// ── TestQualitySelection ──────────────────────────────────────────────────

#[test]
fn test_select_best_quality_resolution_priority() {
    let high_res = ImageMeta {
        path: PathBuf::from("/tmp/high.jpg"),
        width: 3840,
        height: 2160,
        size: 4_000_000,
        mtime: 1000.0,
        sha1: "abc123".to_string(),
        phash: None,
    };

    let low_res = ImageMeta {
        path: PathBuf::from("/tmp/low.jpg"),
        width: 1920,
        height: 1080,
        size: 2_000_000,
        mtime: 2000.0,
        sha1: "def456".to_string(),
        phash: None,
    };

    let candidates = [low_res, high_res.clone()];
    let best = select_best_quality(&candidates).unwrap();
    assert_eq!(best.path, high_res.path);
}

#[test]
fn test_select_best_quality_size_tiebreaker() {
    let file1 = ImageMeta {
        path: PathBuf::from("/tmp/file1.jpg"),
        width: 1920,
        height: 1080,
        size: 3_000_000,
        mtime: 1000.0,
        sha1: "abc123".to_string(),
        phash: None,
    };

    let file2 = ImageMeta {
        path: PathBuf::from("/tmp/file2.jpg"),
        width: 1920,
        height: 1080,
        size: 2_000_000,
        mtime: 1000.0,
        sha1: "def456".to_string(),
        phash: None,
    };

    let candidates = [file1.clone(), file2];
    let best = select_best_quality(&candidates).unwrap();
    assert_eq!(best.path, file1.path);
}

#[test]
fn test_select_best_quality_mtime_tiebreaker() {
    let file1 = ImageMeta {
        path: PathBuf::from("/tmp/file1.jpg"),
        width: 1920,
        height: 1080,
        size: 2_000_000,
        mtime: 2000.0,
        sha1: "abc123".to_string(),
        phash: None,
    };

    let file2 = ImageMeta {
        path: PathBuf::from("/tmp/file2.jpg"),
        width: 1920,
        height: 1080,
        size: 2_000_000,
        mtime: 1000.0,
        sha1: "def456".to_string(),
        phash: None,
    };

    let candidates = [file1.clone(), file2];
    let best = select_best_quality(&candidates).unwrap();
    assert_eq!(best.path, file1.path);
}

// ── TestExactHashGrouping ─────────────────────────────────────────────────

#[test]
fn test_group_by_exact_hash_identical() {
    let meta1 = ImageMeta {
        path: PathBuf::from("/tmp/file1.jpg"),
        width: 1920,
        height: 1080,
        size: 2_000_000,
        mtime: 1000.0,
        sha1: "abc123".to_string(),
        phash: None,
    };

    let meta2 = ImageMeta {
        path: PathBuf::from("/tmp/file2.jpg"),
        width: 1920,
        height: 1080,
        size: 2_000_000,
        mtime: 1000.0,
        sha1: "abc123".to_string(), // Same hash
        phash: None,
    };

    let groups = group_by_exact_hash(&[meta1, meta2]);
    assert_eq!(groups.len(), 1);
    assert_eq!(groups[0].len(), 2);
}

#[test]
fn test_group_by_exact_hash_different() {
    let meta1 = ImageMeta {
        path: PathBuf::from("/tmp/file1.jpg"),
        width: 1920,
        height: 1080,
        size: 2_000_000,
        mtime: 1000.0,
        sha1: "abc123".to_string(),
        phash: None,
    };

    let meta2 = ImageMeta {
        path: PathBuf::from("/tmp/file2.jpg"),
        width: 1920,
        height: 1080,
        size: 2_000_000,
        mtime: 1000.0,
        sha1: "def456".to_string(), // Different hash
        phash: None,
    };

    let groups = group_by_exact_hash(&[meta1, meta2]);
    assert_eq!(groups.len(), 0);
}

#[test]
fn test_group_by_exact_hash_multiple_groups() {
    let items: Vec<ImageMeta> = (0..4)
        .map(|i| ImageMeta {
            path: PathBuf::from(format!("/tmp/file{i}.jpg")),
            width: 100,
            height: 100,
            size: 1000,
            mtime: 1000.0,
            sha1: if i < 2 {
                "hash1".to_string()
            } else {
                "hash2".to_string()
            },
            phash: None,
        })
        .collect();

    let groups = group_by_exact_hash(&items);
    assert_eq!(groups.len(), 2);
    assert!(groups.iter().all(|g| g.len() == 2));
}

// ── TestPerceptualHashGrouping ────────────────────────────────────────────

#[test]
fn test_group_by_perceptual_hash_similar() {
    // Two hashes with small Hamming distance (identical bytes = distance 0).
    let hash_bytes = vec![0xAA, 0xBB, 0xCC, 0xDD];

    let meta1 = ImageMeta {
        path: PathBuf::from("/tmp/file1.jpg"),
        width: 1920,
        height: 1080,
        size: 2_000_000,
        mtime: 1000.0,
        sha1: "abc123".to_string(),
        phash: Some(hash_bytes.clone()),
    };

    let meta2 = ImageMeta {
        path: PathBuf::from("/tmp/file2.jpg"),
        width: 1920,
        height: 1080,
        size: 2_000_000,
        mtime: 1000.0,
        sha1: "def456".to_string(),
        phash: Some(hash_bytes), // Same hash → distance 0
    };

    let groups = group_by_perceptual_hash(&[meta1, meta2], 5, false);
    assert_eq!(groups.len(), 1);
    assert_eq!(groups[0].len(), 2);
}

#[test]
fn test_group_by_perceptual_hash_different() {
    // Two hashes with large Hamming distance.
    let hash1 = vec![0x00, 0x00, 0x00, 0x00];
    let hash2 = vec![0xFF, 0xFF, 0xFF, 0xFF]; // distance = 32

    let meta1 = ImageMeta {
        path: PathBuf::from("/tmp/file1.jpg"),
        width: 1920,
        height: 1080,
        size: 2_000_000,
        mtime: 1000.0,
        sha1: "abc123".to_string(),
        phash: Some(hash1),
    };

    let meta2 = ImageMeta {
        path: PathBuf::from("/tmp/file2.jpg"),
        width: 1920,
        height: 1080,
        size: 2_000_000,
        mtime: 1000.0,
        sha1: "def456".to_string(),
        phash: Some(hash2),
    };

    let groups = group_by_perceptual_hash(&[meta1, meta2], 5, false);
    assert_eq!(groups.len(), 0);
}

#[test]
fn test_group_by_perceptual_hash_aspect_ratio_check() {
    // Same hash but different aspect ratios.
    let hash_bytes = vec![0xAA, 0xBB, 0xCC, 0xDD];

    let meta_wide = ImageMeta {
        path: PathBuf::from("/tmp/wide.jpg"),
        width: 1920,
        height: 1080, // 16:9
        size: 2_000_000,
        mtime: 1000.0,
        sha1: "abc123".to_string(),
        phash: Some(hash_bytes.clone()),
    };

    let meta_square = ImageMeta {
        path: PathBuf::from("/tmp/square.jpg"),
        width: 1000,
        height: 1000, // 1:1
        size: 2_000_000,
        mtime: 1000.0,
        sha1: "def456".to_string(),
        phash: Some(hash_bytes),
    };

    // With aspect ratio check: should NOT group (1.78 vs 1.0).
    let groups =
        group_by_perceptual_hash(&[meta_wide.clone(), meta_square.clone()], 5, true);
    assert_eq!(groups.len(), 0);

    // Without aspect ratio check: should group.
    let groups = group_by_perceptual_hash(&[meta_wide, meta_square], 5, false);
    assert_eq!(groups.len(), 1);
}
